rfcs/0000-unaligned-fn-ptr.md

```diff
@@ -0,0 +1,92 @@
+This is currently an open question.
+
+The motivation to settle this question is that we have to say *something* in the Reference and the Nomicon that is at least a reasonable "upper bound" for the validity requirements of function pointers.
+Otherwise, unsafe code creating function pointers is left without rules to follow.
+With [this PR for the Reference](https://github.com/rust-lang-nursery/reference/pull/659), we have reasonable upper bounds for *most* types of Rust (these have not gone through a formal consensus process, but to my knowledge they form an upper bound for every proposal that has been made for validity of any type thus far).
+The only exception is function pointers.
+Saying anything about their alignment requirements is actually non-trivial: we would have to introduce the notion of a platform-dependent "function alignment", and then ask users to look that up for the platforms they are targeting.
+Preferably, there would be way for code to actually query this alignment requirement (similar to `mem::align_of`).
+To avoid having to create all that infrastructure, I propose we instead declare that function pointers do not have to be aligned.
+
+On most current platforms, we could not make any alignment requirement anyway: on x86, there is no alignment for code, and while ARM does have 2-byte alignment, it uses the least significant bit to encode the instruction set (thumb or not) of the pointee, so function pointers themselves are not guaranteed to be aligned.
+
+Beyond alignment, while we could require the function pointer to point to a "valid sequence of instructions", it is unclear what the benefit of that would be.
+
+Hence, given the lack of a motivation *for* making any non-null function pointers invalid, and in order to simplify the rules unsafe code has to follow and remove needless footguns, we propose to specify null as the only value that is invalid for function pointers.
+
+[ref-ub]: https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html
+[nom-ub]: https://doc.rust-lang.org/nightly/nomicon/what-unsafe-does.html
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When transmuting data to a function pointer, you must make sure it is non-null:
+
+```rust
+fn bad() {
+    let x: fn() = unsafe { mem::transmute(0usize) } // This is UB!
+}
+```
+
+Transmuting any non-null value to a function pointer is defined behavior.
+
+```rust
+fn good() {
+    let x: fn() = unsafe { mem::transmute(1usize) } // This not UB.
+}
+```
+
+Further requirements, outside the scope of this RFC, come into play when the function pointer gets called.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A function pointer is "valid" (in the sense that it can be produced without causing immediate UB) if and only if it is non-null.
+
```

Comment:
Could we add this to the validity chapter of the UCG reference ?  Maybe with the examples from above ?

Reply:
Sure, I can do that once we merged this PR.